File: src/csharp/2_CreationalPatterns/4_Singleton/Singleton.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using Autofac;
using MoreLinq;
using NUnit.Framework;
using static System.Console;

namespace DotNetDesignPatternDemos.Creational.Singleton
{
  public interface IDatabase
  {
    int GetPopulation(string name);
  }

  public class SingletonDatabase : IDatabase
  {
    private Dictionary<string, int> capitals;
    private static int instanceCount;
    public static int Count => instanceCount;

    private SingletonDatabase()
    {
      WriteLine("Initializing database");

      capitals = File.ReadAllLines(
        Path.Combine(
          new FileInfo(typeof(IDatabase).Assembly.Location).DirectoryName, "capitals.txt")
        )
        .Batch(2)
        .ToDictionary(
          list => list.ElementAt(0).Trim(),
          list => int.Parse(list.ElementAt(1)));
    }

    public int GetPopulation(string name)
    {
      return capitals[name];
    }

    // laziness + thread safety
    private static Lazy<SingletonDatabase> instance = new Lazy<SingletonDatabase>(() =>
    {
      instanceCount++;
      return new SingletonDatabase();
    });

    public static IDatabase Instance => instance.Value;
  }

  public class SingletonRecordFinder
  {
    public int TotalPopulation(IEnumerable<string> names)
    {
      int result = 0;
      foreach (var name in names)
        result += SingletonDatabase.Instance.GetPopulation(name);
      return result;
    }
  }

  public class ConfigurableRecordFinder
  {
    private IDatabase database;

    public ConfigurableRecordFinder(IDatabase database)
    {
      this.database = database;
    }

    public int GetTotalPopulation(IEnumerable<string> names)
    {
      int result = 0;
      foreach (var name in names)
        result += database.GetPopulation(name);
      return result;
    }
  }

  public class DummyDatabase : IDatabase
  {
    public int GetPopulation(string name)
    {
      return new Dictionary<string, int>
      {
        ["alpha"] = 1,
        ["beta"] = 2,
        ["gamma"] = 3
      }[name];
    }
  }

  /// <summary>
  /// IMPORTANT: be sure to turn off shadow copying for unit tests in R#!
  /// </summary>
  [TestFixture]
  public class SingletonTests
  {
    [Test]
    public void IsSingletonTest()
    {
      var db = SingletonDatabase.Instance;
      var db2 = SingletonDatabase.Instance;
      Assert.That(db, Is.SameAs(db2));
      Assert.That(SingletonDatabase.Count, Is.EqualTo(1));
    }

    [Test]
    public void SingletonTotalPopulationTest()
    {
      // testing on a live database
      var rf = new SingletonRecordFinder();
      var names = new[] {"Seoul", "Mexico City"};
      int tp = rf.TotalPopulation(names);
      Assert.That(tp, Is.EqualTo(17500000 + 17400000));
    }

    [Test]
    public void DependantTotalPopulationTest()
    {
      var db = new DummyDatabase();
      var rf = new ConfigurableRecordFinder(db);
      Assert.That(
        rf.GetTotalPopulation(new[]{"alpha", "gamma"}),
        Is.EqualTo(4));
    }
  }

  public class Demo
  {
    static void Main()
    {
      var db = SingletonDatabase.Instance;

      // works just fine while you're working with a real database.
      var city = "Tokyo";
      WriteLine($"{city} has population {db.GetPopulation(city)}");

      // now some tests
    }
  }
}

File: src/swift/2_CreationalPatterns/4_Singleton/Singleton.swift
import Foundation
import XCTest

protocol Database
{
  func getPopulation(name: String) -> Int
}

class SingletonDatabase : Database
{
  var capitals = [String: Int]()
  static var instanceCount = 0
  static var instance = SingletonDatabase()

  private init()
  {
    type(of: self).instanceCount += 1
    print("Initializing database")

    let path = "/mnt/c/Dropbox/Projects/Demos/SwiftDesignPatterns/patterns/creational/singleton/capitals.txt"
    if let text = try? String(contentsOfFile: path as String, 
      encoding: String.Encoding.utf8)
    {
      let strings = text.components(separatedBy: .newlines)
        .filter { !$0.isEmpty }
        .map { $0.trimmingCharacters(in: .whitespaces)}
      //print(strings.count)
      for i in 0..<(strings.count/2)
      {
        //print("`\(strings[i*2])` has population \(Int(strings[i*2+1])!)")
        capitals[strings[i*2]] = Int(strings[i*2+1])!
      }
    }
  }

  func getPopulation(name: String) -> Int {
    return capitals[name]!
  }
}

class SingletonRecordFinder {
  func totalPopulation(names: [String]) -> Int
  {
    var result = 0
    for name in names {
      // singleton database hardcoded here
      result += SingletonDatabase.instance.getPopulation(name: name);
    }
    return result
  }
}

class ConfigurableRecordFinder {
  let database: Database
  init(database: Database)
  {
    self.database = database
  }
  func totalPopulation(names: [String]) -> Int
  {
    var result = 0
    for name in names {
      result += database.getPopulation(name: name);
    }
    return result
  }
}

class DummyDatabase : Database
{
    func getPopulation(name: String) -> Int {
      return ["alpha": 1, "beta": 2, "gamma": 3][name]!
    }
}

class SingletonTests: XCTestCase
{
  static var allTests = [
    ("test_isSingletonTest", test_isSingletonTest),
    ("test_singletonTotalPopulationTest", test_singletonTotalPopulationTest),
    ("test_dependantTotalPopulationTest", test_dependantTotalPopulationTest)
  ]

  func test_isSingletonTest()
  {
    var db = SingletonDatabase.instance
    var db2 = SingletonDatabase.instance
    XCTAssertEqual(1, SingletonDatabase.instanceCount, "instance count must = 1")
  }

  func test_singletonTotalPopulationTest()
  {
    let rf = SingletonRecordFinder()
    let names = ["Seoul", "Mexico City"]
    let tp = rf.totalPopulation(names: names)
    XCTAssertEqual(17_500_000+17_400_000, tp, "population size must match")
  }

  func test_dependantTotalPopulationTest() {
    let db = DummyDatabase()
    let rf = ConfigurableRecordFinder(database: db)
    XCTAssertEqual(4, rf.totalPopulation(names: ["alpha", "gamma"]))
  }
}

func main()
{
  XCTMain([testCase(SingletonTests.allTests)])
}

func main_old()
{
  // cannot construct database directly
  //let mydb = SingletonDatabase()

  let db = SingletonDatabase.instance
  var city = "Tokyo"
  print("\(city) has population \(db.getPopulation(name: city))")
  city = "Manila"
  print("\(city) has population \(db.getPopulation(name: city))")
  print("SingletonDatabase has \(SingletonDatabase.instanceCount) instance(s)")
}

main()